cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -867,7 +867,8 @@ Result<std::shared_ptr<DatasetFactory>> ParquetDatasetFactory::Make(
   ARROW_ASSIGN_OR_RAISE(auto physical_schema, GetSchema(*metadata, properties));
   ARROW_ASSIGN_OR_RAISE(auto manifest, GetSchemaManifest(*metadata, properties));
 
-  std::unordered_map<std::string, std::vector<int>> path_to_row_group_ids;
+  std::vector<std::pair<std::string, std::vector<int>>> paths_with_row_group_ids;
+  std::unordered_map<std::string, int> paths_to_index;
```

Comment:
FTR, I'll mention `std::unordered_multimap`

Reply:
I always forget about multimaps :).  I'm going to keep this as-is for now but I will mentally jot this down to improve the next time I get a chance.